src/main/java/com/wanderersoftherift/wotr/block/RiftSpawnerBlock.java

```diff
@@ -0,0 +1,113 @@
+                    Direction.WEST,
+                    Block.box(3.0, 0.0, 0.0, 16.0, 16.0, 16.0),
+                    Direction.EAST,
+                    Block.box(0.0, 0.0, 0.0, 13.0, 16.0, 16.0)
+            )
+    );
+
+
+    public static final EnumProperty<Direction> FACING = BlockStateProperties.FACING;
+
+    public RiftSpawnerBlock(Properties properties) {
+        super(properties);
+        registerDefaultState(stateDefinition.any().setValue(FACING, Direction.UP));
+    }
+
+    public @NotNull MapCodec<RiftSpawnerBlock> codec() {
+        return CODEC;
+    }
+
+    @Override
+    protected @NotNull VoxelShape getShape(BlockState state, @NotNull BlockGetter level, @NotNull BlockPos pos, @NotNull CollisionContext context) {
+        return SHAPES.get(state.getValue(BlockStateProperties.FACING));
+    }
+
+    @Override
+    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
+        builder.add(FACING);
+    }
+
+    @Override
+    public @Nullable BlockState getStateForPlacement(BlockPlaceContext context) {
+        return defaultBlockState().setValue(FACING, context.getClickedFace());
+    }
+
+    /**
+     * Provides the location to create the rift, if a valid location exists
+     * @param level
+     * @param pos
+     * @param dir
+     * @return A valid spawn location, or Optional#empty
+     */
+    public Optional<SpawnLocation> getSpawnLocation(Level level, BlockPos pos, Direction dir) {
+        Direction facing = level.getBlockState(pos).getValue(BlockStateProperties.FACING);
+        if (facing.getAxis().isVertical()) {
+            BlockPos checkPos = pos;
+            for (int i = 0; i < 3; i++) {
+                checkPos = checkPos.relative(facing);
+                if (!allowsPortal(level, checkPos)) {
+                    return Optional.empty();
+                }
+            }
+            if (facing == Direction.UP) {
+                return Optional.of(new SpawnLocation(pos.above(), pos.above().getBottomCenter(), Direction.UP));
+            } else {
+                BlockPos origin = pos.relative(dir, 3);
+                return Optional.of(new SpawnLocation(origin, origin.getBottomCenter(), Direction.UP));
+            }
+        }
+
+
+        BlockPos adjacentPos = pos.relative(facing);
+        if (allowsPortal(level, adjacentPos) && allowsPortal(level, adjacentPos.above()) && allowsPortal(level, adjacentPos.below())) {
+            return Optional.of(new SpawnLocation(adjacentPos, adjacentPos.below().getBottomCenter().subtract(facing.getUnitVec3().scale(0.475)), facing));
+        }
+        return Optional.empty();
+    }
+
+    private boolean allowsPortal(Level level, BlockPos pos) {
+        BlockState state = level.getBlockState(pos);
+        return state.isAir() || state.getCollisionShape(level, pos).isEmpty();
+    }
+
+    public record SpawnLocation(BlockPos blockPos, Vec3 position, Direction direction) {}
```

Comment:
I personally do not like public classes inside another class. If it is public, it is best to move it to its own file.

Reply:
Done, moved to `com.wanderersoftherift.wotr.entity.portal.PortalSpawnLocation`